elfo-core/src/message.rs

```diff
@@ -26,439 +33,126 @@ pub trait Message: fmt::Debug + Clone + Any + Send + Serialize + for<'de> Deseri
         self._vtable().protocol
     }
 
-    #[doc(hidden)]
+    #[doc(hidden)] // unstable because depends on `metrics`
     #[inline(always)]
     fn labels(&self) -> &'static [Label] {
-        self._vtable().labels
+        &self._vtable().labels
     }
 
-    #[doc(hidden)]
+    #[doc(hidden)] // unstable because will be replaced with `DumpingMode`
     #[inline(always)]
     fn dumping_allowed(&self) -> bool {
         self._vtable().dumping_allowed
     }
 
+    #[deprecated(note = "use `AnyMessage::new` instead")]
     #[doc(hidden)]
     #[inline(always)]
     fn upcast(self) -> AnyMessage {
-        self._touch();
-        AnyMessage {
-            vtable: self._vtable(),
-            data: smallbox!(self),
-        }
+        self._into_any()
     }
 
     // Private API.
 
     #[doc(hidden)]
-    fn _vtable(&self) -> &'static MessageVTable;
+    fn _type_id() -> MessageTypeId;
 
-    // Called while upcasting/downcasting to avoid
-    // [rust#47384](https://github.com/rust-lang/rust/issues/47384).
     #[doc(hidden)]
-    fn _touch(&self);
+    fn _vtable(&self) -> &'static MessageVTable;
 
     #[doc(hidden)]
     #[inline(always)]
-    fn _erase(&self) -> dumping::ErasedMessage {
-        smallbox!(self.clone())
+    fn _repr_layout(&self) -> alloc::Layout {
+        self._vtable().repr_layout
     }
-}
 
-pub trait Request: Message {
-    type Response: fmt::Debug + Clone + Send + Serialize;
+    // NOTE: All methods below MUST be overriden for `AnyMessage`.
 
     #[doc(hidden)]
-    type Wrapper: Message + Into<Self::Response> + From<Self::Response>;
-}
-
-// === AnyMessage ===
-
-// Reexported in `elfo::_priv`.
-pub struct AnyMessage {
-    vtable: &'static MessageVTable,
-    data: SmallBox<dyn Any + Send, [usize; 24]>,
-}
-
-impl AnyMessage {
-    #[inline]
-    pub fn is<M: Message>(&self) -> bool {
-        self.data.is::<M>()
-    }
-
-    #[inline]
-    pub fn type_id(&self) -> std::any::TypeId {
-        (*self.data).type_id()
-    }
-
-    #[inline]
-    pub fn downcast_ref<M: Message>(&self) -> Option<&M> {
-        self.data.downcast_ref::<M>().map(|message| {
-            message._touch();
-            message
-        })
+    #[inline(always)]
+    fn _is_supertype_of(type_id: MessageTypeId) -> bool {
+        Self::_type_id() == type_id
     }
 
-    #[inline]
-    pub fn downcast<M: Message>(self) -> Result<M, AnyMessage> {
-        if !self.is::<M>() {
-            return Err(self);
-        }
-
-        let message = self
-            .data
-            .downcast::<M>()
-            .expect("cannot downcast")
-            .into_inner();
-
-        message._touch();
-        Ok(message)
-    }
-}
+    // Called in `_read()` and `_write()` to avoid
```

Comment:
I’m not sure what’s the problem `._touch()` is trying to avoid. Linked issues are related to `#[no_mangle]` / `#[used]` statics, but this is a method that’s called from other methods?

Reply:
We hit this problem two years ago in the case of protocol crates without any callable code, only message declarations. The problem was that the compiler ignored `#[used]` in that case, and I'm not sure it still happens now.

Added `#[inline(never)]` callable empty `touch()` fixed this problem.

Reply:
So the problem is solved by calling any not-inlined function from the crate containing `#[used]` static? I don’t really understand why it’s `._read()` / `._write()` methods that do it.